### PR TITLE
speedreader: Completely separate states

### PR DESCRIPTION
**Pull request by keur**
* Do not conflate Reader Mode state and Speedreader state. This leads
    to confusion.
  * Due to removing the conflation, delete the dependency on the
    Chromium distillability observer. We now use our own URL heuristic.
  * Have SpeedreaderTabHelper::UpdateActiveState() manage all
    Speedreader state for a stage. The bubbles and icons will just check
    that variable. Vastly simplifies logic.
  * Demarcate label and icon colors when Speedreader is enabled or
    disabled for a page.

As a bonus, the code should be much easier to understand now.

Resolves https://github.com/brave/brave-browser/issues/16663

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/speedreader/speedreader_pref_names.h [hunk 1]
```diff
@@ -8,10 +8,18 @@
...
+// Is Speedreader currently enabled
 constexpr char kSpeedreaderPrefEnabled[] = "brave.speedreader.enabled";
+
+// Set if Speedreader was enabled at least once
 constexpr char kSpeedreaderPrefEverEnabled[] = "brave.speedreader.ever_enabled";
+
+// Number of times the user has toggled Speedreader
 constexpr char kSpeedreaderPrefToggleCount[] = "brave.speedreader.toggle_count";
 
+// Number of times the "Enable Speedreader" button was show automatically
```
Comment (iefremov) [inline components/speedreader/speedreader_pref_names.h:20]: was shown ?

**Comment by iefremov** [comment]
I've found a state when the speedreader button is white on one tab and blue on another (same domain for two tabs), when disabled. Is it expected? 
 1. open any article on lenta.ru
 2. disable speedreader via the bubble
 3. open any other article in a new tab
 

https://user-images.githubusercontent.com/3109556/124498545-bb541b00-dde6-11eb-9b5b-a35256ab5133.mov

File: browser/ui/browser_commands.cc [hunk 2]
```diff
@@ -84,50 +84,35 @@ void OpenGuestProfile() {
...
+        break;
+      case DistillState::kPageProbablyReadable: {
+        auto* speedreader_service =
+            speedreader::SpeedreaderServiceFactory::GetForProfile(
+                browser->profile());
         tab_helper->SingleShotSpeedreader();
-        if (!speedreader_enabled)
+        if (speedreader_service->ShouldPromptUserToEnable()) {
           tab_helper->ShowReaderModeBubble();
+          speedreader_service->IncrementPromptCount();
```
Comment (iefremov) [inline browser/ui/browser_commands.cc:110]: I was thinking that maybe we can move this logic from `browser_commands.cc`? seems too complex to sit there

**Comment by iefremov** [comment]
Do we have an updated spec or something like a state diagram?

File: browser/ui/browser_commands.cc (hunk 2 above)
Comment (keur) [reply browser/ui/browser_commands.cc]: fixed

**Comment by keur** [comment]
@iefremov Fixed the UI bug you found. I put an explanation in the commit message. We don't have a UI block diagram, only the spec @karenkliu put on Figma for us.

File: browser/speedreader/speedreader_tab_helper.h [hunk 3]
```diff
@@ -63,6 +63,7 @@ class SpeedreaderTabHelper
   // Returns |true| if the user has enabled Speedreader but the domain in the
   // active web contents is blacklisted.
   bool IsEnabledForSite();
+  bool IsEnabledForSite(const GURL& url);
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.h:66]: maybe we can make it private if nobody except the tab helper wants it

**Comment by iefremov** [comment]
> @iefremov Fixed the UI bug you found. I put an explanation in the commit message. We don't have a UI block diagram, only the spec @karenkliu put on Figma for us.

Yes, I'm just curious whether this impl is aligned with what we have on Figma?

**Review by iefremov — APPROVED** [review]

**Comment by keur** [comment]
> Yes, I'm just curious whether this impl is aligned with what we have on Figma?

Yeah, these changes are direct follow-ups to meetings with @karenkliu and @rebron. They tried it out, told me what to change to be more in line with the spec.

**Comment by keur** [comment]
CI already passed for everything, just moved a method to private and changed typo in comment.

File: components/speedreader/speedreader_pref_names.h (hunk 1 above)
Comment (keur) [reply components/speedreader/speedreader_pref_names.h]: fixed.

File: browser/speedreader/speedreader_tab_helper.h (hunk 3 above)
Comment (keur) [reply browser/speedreader/speedreader_tab_helper.h]: fixed

**Pull request merged**